yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnShuffleEncryptionSuite.scala

```diff
@@ -0,0 +1,226 @@
+import org.mockito.Mockito.when
+import org.scalatest.{BeforeAndAfterAll, Matchers}
+
+import org.apache.spark._
+import org.apache.spark.crypto.{CryptoConf, CryptoStreamUtils}
+import org.apache.spark.deploy.SparkHadoopUtil
+import org.apache.spark.executor.{ShuffleWriteMetrics, TaskMetrics}
+import org.apache.spark.io.CompressionCodec
+import org.apache.spark.memory.{TestMemoryManager, TaskMemoryManager}
+import org.apache.spark.network.util.LimitedInputStream
+import org.apache.spark.serializer.{KryoSerializer, SerializerInstance}
+import org.apache.spark.shuffle.IndexShuffleBlockResolver
+import org.apache.spark.shuffle.sort.{SerializedShuffleHandle, UnsafeShuffleWriter}
+import org.apache.spark.storage.{BlockId, BlockManager, DiskBlockManager, DiskBlockObjectWriter,
+  TempShuffleBlockId}
+import org.apache.spark.util.Utils
+
+class YarnShuffleEncryptionSuite extends SparkFunSuite with Matchers with BeforeAndAfterAll{
+  val NUM_PARTITITONS = 4
+  val conf = new SparkConf()
+  val tempDir = Utils.createTempDir("test", "test")
+  val mergedOutputFile = File.createTempFile("mergedoutput", "", tempDir)
+  val memoryManager = new TestMemoryManager(conf)
+  val taskMemoryManager = new TaskMemoryManager(memoryManager, 0)
+  val serializer = new KryoSerializer(new SparkConf())
+  val hashPartitioner = new HashPartitioner(NUM_PARTITITONS)
+  val taskMetrics = new TaskMetrics()
+  val spillFilesCreated = new LinkedList[File]()
+
+  var partitionSizesInMergedFile: Array[Long] = null
+  val ugi = UserGroupInformation.createUserForTesting("testuser", Array[String] {
+    "testgroup"
+  })
+
+  @Mock(answer = RETURNS_SMART_NULLS) private var blockManager: BlockManager = _
+  @Mock(answer = RETURNS_SMART_NULLS) private var diskBlockManager: DiskBlockManager = _
+  @Mock(answer = RETURNS_SMART_NULLS) private var taskContext: TaskContext = _
+  @Mock(answer = RETURNS_SMART_NULLS) private var blockResolver: IndexShuffleBlockResolver = _
+  @Mock(
+    answer = RETURNS_SMART_NULLS) private var shuffleDep: ShuffleDependency[Object, Object,
+      Object] = _
+
+  override def beforeAll(): Unit = {
+    MockitoAnnotations.initMocks(this)
+    System.setProperty("SPARK_YARN_MODE", "true")
+    ugi.doAs(new PrivilegedExceptionAction[Unit]() {
+      override def run(): Unit = {
+        initialShuffleWriter()
+      }
+    })
+  }
+
+  override def afterAll(): Unit = {
+    Utils.deleteRecursively(tempDir);
+    val leakedMemory = taskMemoryManager.cleanUpAllAllocatedMemory();
+    if (leakedMemory != 0) {
+      fail("Test leaked " + leakedMemory + " bytes of managed memory");
+    }
+  }
+
+  test("Test Yarn shuffle encryption basic write") {
+    val doAsAction = new PrivilegedExceptionAction[Unit] {
+      override def run(): Unit = {
+        testYarnShuffleEncryptionWrite()
+      }
+    }
+    ugi.doAs(doAsAction)
+  }
+
+  private[this] def testYarnShuffleEncryptionWrite(): Unit = {
+    val dataToWrite = new ArrayList[Product2[Object, Object]]()
```

Comment:
Does this work if you use:

    new ArrayList[(Int, Int)]

That would clean up some of the ugly casting in the code.